extensions/cornerstone/src/services/SegmentationService/SegmentationService.ts

```diff
@@ -719,13 +721,162 @@ class SegmentationService {
+          });
+
+          return {
+            ...rest,
+            points: newPoints,
+          };
+        });
+
+        return {
+          data,
+          id: ROIName || ROINumber,
+          segmentIndex: ROINumber,
+          color: colorArray,
+          geometryId: `${rtDisplaySetUID}:${ROIName ||
+            ROINumber}:segmentIndex-${ROINumber}`,
+        };
+      }
+    );
+
+    // sort by segmentIndex
+    allRTStructData.sort((a, b) => a.segmentIndex - b.segmentIndex);
+
+    const geometryIds = allRTStructData.map(({ geometryId }) => geometryId);
+
+    const segmentation: Segmentation = {
+      ...defaultScheme,
+      id: segmentationId,
+      displaySetInstanceUID: rtDisplaySetUID,
+      type: representationType,
+      representationData: {
+        [CONTOUR]: {
+          geometryIds,
+        },
+      },
+    };
+
+    const cachedSegmentation = this.getSegmentation(segmentationId);
+
+    if (cachedSegmentation) {
+      // if the labelmap with the same segmentationId already exists, we can
+      // just assume that the segmentation is already created and move on with
+      // updating the state
+      return this.addOrUpdateSegmentation(
+        Object.assign(segmentation, cachedSegmentation),
+        suppressEvents
+      );
+    }
+
+    if (!structureSet.ROIContours?.length) {
+      throw new Error(
+        'The structureSet does not contain any ROIContours. Please ensure the structureSet is loaded first.'
+      );
+    }
+    let count = 0;
+    const segmentsCachedStats = {};
+    const initializeContour = async rtStructData => {
+      const { data, id, color, segmentIndex, geometryId } = rtStructData;
+      const geometry = await geometryLoader.createAndCacheGeometry(geometryId, {
+        geometryData: {
+          data,
+          id,
+          color,
+          frameOfReferenceUID: structureSet.frameOfReferenceUID,
+          segmentIndex,
+        },
+        type: csEnums.GeometryType.CONTOUR,
+      });
+
+      const contourSet = geometry.data;
+      const centroid = contourSet.getCentroid();
+
+      segmentsCachedStats[segmentIndex] = {
+        center: { world: centroid },
+        modifiedTime: Date.now(),
+      };
+
+      segmentation.segments[segmentIndex] = {
+        label: id,
+        segmentIndex,
+        color,
+        opacity: 255,
+        isVisible: true,
+        isLocked: false,
+      };
+
+      this._broadcastEvent(EVENTS.SEGMENT_LOADING_COMPLETE, {
+        segmentIndex: count++,
+        // Note: this is not the geometryIds length since there might be
+        // some missing ROINumbers
+        numSegments: allRTStructData.length,
+      });
+    };
+
+    for (let i = 0; i < allRTStructData.length; i++) {
+      const promise = new Promise<void>((resolve, reject) => {
+        setTimeout(() => {
+          initializeContour(allRTStructData[i]).then(() => {
+            resolve();
+          });
+        }, 0);
+      });
+
+      await promise;
```

Comment:
This for loop and await within it appears to suggest that only one contour can be initialized at time. If that is true, then this is fine as is, but otherwise consider using something like creating an array of `Promise`s and using `await Promise.all` for them to complete.